Refactoring, Dynamic prefix and AWS v2

ph opened the pull request:
**Motivation**
One of the most requested features was adding a way to add dynamic prefixes using the fieldref
syntax for the files on the bucket and also the changes in the pipeline to support shared delegator.
The S3 output by nature was always a single threaded writes but had multiples workers to process the upload, the code was threadsafe when used in the concurrency `:single` mode.

This PR addresses a few problems and provide shorter and more structured code:
- This Plugin now uses the V2 version of the SDK, this make sure we receive the latest updates and changes.
- We now uses S3's `upload_file` instead of reading chunks, this method is more efficient and will uses the multipart with threads if the files is too big.
- You can now use the `fieldref` syntax in the prefix to dynamically changes the target with the events it receives.
- The Upload queue is now a bounded list, this options is necessary to allow back pressure to be communicated back to the pipeline but its configurable by the user.
- If the queue is full the plugin will start the upload in the current thread.
- The plugin now threadsafe and support the concurrency model `shared`
- The rotation strategy can be selected, the recommended is `size_and_time` that will check for both the configured limits (`size` and `time` are also available)
- The `restore` option will now use a separate threadpool with an unbounded queue
- The `restore` option will not block the launch of logstash and will uses less resources than the real time path
- The plugin now uses `multi_receive_encode`, this will optimize the writes to the files
- rotate operation are now batched to reduce the number of IO calls.
- Empty file will not be uploaded by any rotation rotation strategy
- We now use Concurrent-Ruby for the implementation of the java executor
- If you have finer grain permission on prefixes or want faster boot, you can disable the credentials check with `validate_credentials_on_root_bucket`
- The credentials check will no longer fails if we can't delete the file
- We now have a full suite of integration test for all the defined rotation

Fixes: #4 #81 #44 #59 #50

tkampour commented:
How likely is this to be merged soon?

ph commented:
@tkampour Since this is a big change, in fact a complete rewrite, it need to go through the review process like any other PRs. Since you have commented on this issue you will get notified when it happen, since its this one of most requested feature I am also eager to get this merged in.

lremurphy commented:
Awesome work!  Can't wait to try this!

vistorve commented:
I too am very excited for this change, so excited I am attempting to create a local build of it to get working but I keep running into the following:

Couldn't find any output plugin named 's3'. Are you sure this is correct? Trying to load the s3 output plugin resulted in this error: no such file to load -- concurrent/map", :level=>:error, :file=>"logstash/agent.rb", :line=>"448", :method=>"create_pipeline"}
{:timestamp=>"2016-09-20T16:29:51.292000+0000", :message=>"starting agent", :level=>:info, :file=>"logstash/agent.rb", :line=>"213", :method=>"execute"}

I built the gem and am using that with --no-verify to install. Logstash version=2.4, I am doing this inside a docker if that makes a difference.

suyograo commented:
@vistorve how did you build this gem?  are you following the steps here? https://github.com/logstash-plugins/logstash-output-s3#2-running-your-unpublished-plugin-in-logstash

vistorve commented:
I did the steps in 2.2

ph commented:
@vistorve Nice finding, this is a bug in this PR. I am using **concurrent-ruby**'s map implementation and its available in 1.0.0 >= Logstash 5.0/master ships with it but 2.4 ships with 0.9.2. The dependency is not strict enough in the gemspec this is why you were able to install it but failed to run.

logstash-core depends on a strict version of it in https://github.com/elastic/logstash/blob/2.4/Gemfile.jruby-1.9.lock#L162 so if the plugin requires a newer version the core wont permit it.

@suyograo If we want to make this change works in both 2.4 and 5.0 we have the following solutions:

1. When running in 2.4 Using `threadsafe::hash` that already shipped with logstash, otherwise we use concurrent-ruby's map.

2. Replace concurrent-ruby's map with https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentMap.html which is part of the standard library.

3. release a 2.4.1 with a concurrent-ruby updated.

The other features I used from concurrent-ruby ScheduledTask and the threadpool are available in 0.9.2.

suyograo commented:
@ph can we do #2 since it looks like a safe option across 2.4 and 5.0?

ph commented:
@suyograo #2? We have the `temporary_directory` option for quite a while now?

suyograo commented:
@ph I meant point 2:

> Replace concurrent-ruby's map with https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentMap.html which is part of the standard library.

ph commented:
/facepalm ☕ ☕

andrewvc commented:
FWIW I generally prefer the java util concurrent stuff to the concurrent ruby objects. IMHO the main point of concurrent ruby is to be portable across ruby implementations' concurrency models, which is not a concern for Logstash.

andrewvc commented on lib/logstash/outputs/s3/size_and_time_rotation_policy.rb:
```diff
@@ -0,0 +1,24 @@
+          @size_strategy.rotate?(file) || @time_strategy.rotate?(file)
+        end 
+
+        def need_periodic?
```
s/need_periodic/needs_periodic/

andrewvc commented on lib/logstash/outputs/s3/temporary_file.rb:
```diff
@@ -0,0 +1,45 @@
+        extend Forwardable
+        DELEGATES_METHODS = [:path, :write, :close, :size, :fsync]
+
+        def_delegators :@fd, *DELEGATES_METHODS
```
AFAICT we only use a constant here to test if these methods are defined in a later spec. I would just define them inline and remove the spec. I don't think there's a ton of value in specs that check that methods merely exist since at that point we're testing that `Forwardable` isn't broken. 

I don't think we need to do any testing here FWIW and we can simplify this.

andrewvc commented on lib/logstash/outputs/s3/temporary_file_factory.rb:
```diff
@@ -0,0 +1,90 @@
+        end
+
+        def generate_name
+          filename = "ls.s3.#{Socket.gethostname}.#{current_time}"
```
We need to document that there is a (remote) possibility of collisions between LS instances if this scheme is used. If users use multiple logstashes they should probably give them unique prefixes.

Alternatively we could inject some randomness into here as well.

ph replied:
what do are you suggesting for randomness? 

We have:
- the host
- a date

SecureRandom.uuid?

ph commented on lib/logstash/outputs/s3.rb:
```diff
@@ -63,37 +65,54 @@
   # If you have tags then it will generate a specific size file for every tags
   ##NOTE: define size of file is the better thing, because generate a local temporary file on disk and then put it in bucket.
-  config :size_file, :validate => :number, :default => 0
+  config :size_file, :validate => :number, :default => 1024 * 1024 * 5
```
+1, I wanted to change theses names for quite a while, I will keep the other working and mark them as deprecated.

ph commented on lib/logstash/outputs/s3/temporary_file_factory.rb:
```diff
@@ -0,0 +1,90 @@
+      #
+      # Since the UUID should be fairly unique I can destroy the whole path when an upload is complete.
+      # I do not have to mess around to check if the other directory have file in it before destroying them.
+      class TemporaryFileFactory
```
I agree it feel more like a proxy to me, in this case.

s4nch3z commented on lib/logstash/outputs/s3/time_rotation_policy.rb:
```diff
@@ -0,0 +1,26 @@
+        end
+
+        def rotate?(file)
+          file.size > 0 && Time.now - file.ctime >= time_file
```
Docs says ```time_file``` is in MINUTES, but it's not like that here

nevins-b commented:
There hasn't seemed to be any activity on this pull request in a while so I made a [fork](/nevins-b/logstash-output-s3/tree/feature/cleaner-code) and modified the code to address most of the issues raised with this review

ph commented:
@nevins Thanks can you make a PR with your changes? I will help moving it forward :)

nevins-b commented:
@ph thanks! opened #105

ph commented:
I seem to run into a weird jruby thread/autoload issue when running the tests. To be clear the bug was there before but the new logger infra expose it.


https://github.com/jruby/jruby/issues/3920

ph commented:
@andrewvc This PR has been updated with your requested changes:

- Fix an issue with gzip encoding, its wasn't working, I have added 2 integration test to make sure the behavior is fine.
- Added the code from @nevins-b 
- making sure all the unit test/integration test pass

I believe its ready for the final review.

andrewvc commented:
Got this error testing with a simple config:

```
[2016-12-07T13:35:33,968][ERROR][logstash.outputs.s3      ] Uploading failed, retrying {:exception=>#<Aws::S3::Errors::MalformedXML: The XML you provided was not well-formed or did not validate against our published schema>, :path=>"/var/folders/x1/v84r4j5j4kx3yvyqrlnhst300000gn/T/logstash/ls.s3.andrewvc-elastic.2016-12-07T12.32.part0.txt", :backtrace=>["/Users/andrewvc/Downloads/logstash-5.0.2/vendor/bundle/jruby/1.9/gems/aws-sdk-core-2.3.22/lib/seahorse/client/plugins/raise_response_errors.rb:15:in `call'", "/Users/andrewvc/Downloads/logstash-5.0.2/vendor/bundle/jruby/1.9/gems/aws-sdk-core-2.3.22/lib/aws-sdk-core/plugins/s3_sse_cpk.rb:19:in `call'", "/Users/andrewvc/Downloads/logstash-5.0.2/vendor/bundle/jruby/1.9/gems/aws-sdk-core-2.3.22/lib/aws-sdk-core/plugins/s3_accelerate.rb:33:in `call'", "/Users/andrewvc/Downloads/logstash-5.0.2/vendor/bundle/jruby/1.9/gems/aws-sdk-core-2.3.22/lib/aws-sdk-core/plugins/param_converter.rb:20:in `call'", "/Users/andrewvc/Downloads/logstash-5.0.2/vendor/bundle/jruby/1.9/gems/aws-sdk-core-2.3.22/lib/seahorse/client/plugins/response_target.rb:21:in `call'", "/Users/andrewvc/Downloads/logstash-5.0.2/vendor/bundle/jruby/1.9/gems/aws-sdk-core-2.3.22/lib/seahorse/client/request.rb:70:in `send_request'", "/Users/andrewvc/Downloads/logstash-5.0.2/vendor/bundle/jruby/1.9/gems/aws-sdk-core-2.3.22/lib/seahorse/client/base.rb:207:in `put_object'", "/Users/andrewvc/Downloads/logstash-5.0.2/vendor/bundle/jruby/1.9/gems/aws-sdk-resources-2.3.22/lib/aws-sdk-resources/services/s3/file_uploader.rb:42:in `put_object'", "/Users/andrewvc/Downloads/logstash-5.0.2/vendor/bundle/jruby/1.9/gems/aws-sdk-resources-2.3.22/lib/aws-sdk-resources/services/s3/file_uploader.rb:49:in `open_file'", "/Users/andrewvc/Downloads/logstash-5.0.2/vendor/bundle/jruby/1.9/gems/aws-sdk-resources-2.3.22/lib/aws-sdk-resources/services/s3/file_uploader.rb:41:in `put_object'", "/Users/andrewvc/Downloads/logstash-5.0.2/vendor/bundle/jruby/1.9/gems/aws-sdk-resources-2.3.22/lib/aws-sdk-resources/services/s3/file_uploader.rb:34:in `upload'", "/Users/andrewvc/Downloads/logstash-5.0.2/vendor/bundle/jruby/1.9/gems/aws-sdk-resources-2.3.22/lib/aws-sdk-resources/services/s3/object.rb:251:in `upload_file'", "/Users/andrewvc/projects/lsp/logstash-output-s3/lib/logstash/outputs/s3/uploader.rb:38:in `upload'", "/Users/andrewvc/projects/lsp/logstash-output-s3/lib/logstash/outputs/s3/uploader.rb:29:in `upload_async'", "org/jruby/RubyProc.java:281:in `call'", "/Users/andrewvc/Downloads/logstash-5.0.2/vendor/bundle/jruby/1.9/gems/concurrent-ruby-1.0.0-java/lib/concurrent/executor/java_executor_service.rb:94:in `run'", "Concurrent$$JavaExecutorService$$Job_1607972051.gen:13:in `run'"]}
```

Config:

```

input {
  generator  { count => 100}
}

filter {
}

output {
  s3 {
    access_key_id => "REDACTED"
    secret_access_key => "REDACTED"
    bucket => "REDACTED"
    size_file => 1048576 # 1MB
    region => "us-east-1"
    canned_acl => public_read
  }

  stdout {
    codec => dots
  }
}
```

ph commented:
@andrewvc the isses seems to be resolved now, I haven't been able to reproduce it on my side.
We have zoomed on this to dig deeper in the aws SDK framework. Maybe it could have been a temporary hiccups we don't know for sure at this time.

andrewvc commented:
@ph have you confirmed that the 'gzip' encoding works correctly? When I enable it I don't get gzipped content on S3.

ph commented:
@andrewvc I've tried it this morning with gzip encoding with the included config, what is happening to you? No files at all? Do you use a fixed amount of events? If you are using the 1mb `size_file` limit it can take a while to generate a gzip file.

If I take a 1mb gzip file and uncompress it, I get a 89 mb text file that contains 2132597 events.

```
input { generator {} }
filter {}
output {
    s3 {
        access_key_id => "redacted"
            secret_access_key => "redacted"
            bucket => "redacte"
            region => "us-east-1"
            encoding => "gzip"
            size_file => 1048576
            canned_acl => public_read
    }
    stdout {
        codec => dots
    }
}
```

andrewvc commented:
@ph the problem was that the files were uploaded plain encoded :(

andrewvc commented:
@ph my mistake, it was working right.

andrewvc commented on lib/logstash/outputs/s3/temporary_file_factory.rb:
```diff
@@ -0,0 +1,122 @@
+        end
+
+        class IOWrappedGzip
+          extend Forwardable
```
Can you `require 'forwardable'` at the top of this file?

andrewvc commented:
Left a few more comments, it's almost good to go!

ph commented:
@andrewvc I have addressed your latest comment.

ph commented:
rekicking the travis job

andrewvc commented:
@ph approved! Great work :)
